packages/console/src/hooks/use-connector-groups.ts

```diff
@@ -0,0 +1,54 @@
+import { ConnectorDTO } from '@logto/schemas';
+import { useMemo } from 'react';
+import useSWR from 'swr';
+
+import { RequestError } from '@/hooks/use-api';
+import { ConnectorGroup } from '@/types/connector';
+
+// Group connectors by target
+const useConnectorGroups = () => {
+  const { data, ...rest } = useSWR<ConnectorDTO[], RequestError>('/api/connectors');
+
+  const groups = useMemo(() => {
+    if (!data) {
+      return;
+    }
+
+    return data.reduce<ConnectorGroup[]>((previous, item) => {
+      const groupIndex = previous.findIndex(({ target }) => target === item.target);
+
+      if (groupIndex === -1) {
+        return [
+          ...previous,
+          {
+            name: item.metadata.name,
+            logo: item.metadata.logo,
+            target: item.metadata.target,
+            enabled: item.enabled,
+            connectors: [item],
+          },
+        ];
+      }
+
+      return previous.map((group, index) => {
+        if (index !== groupIndex) {
+          return group;
+        }
```

Comment:
will this ever happen?

Reply:
YES, only push new connectors to the specific index.